### ISSUE TITLE
Model for viral escape

### ISSUE THREAD

**Issue by timcyu** [matsengrp/torchdms#94]
The idea is to create a model that can learn the effect of each mutation on binding at every epitope targeted by polyclonal sera antibodies. Lets say there are `E` epitopes and `S` amino acid sites on the protein.

## General features
1. **epitope weights**: A binary tensor of size (`E`,`S`). If site `s` in `S` is located in site `e` in `E`, then 1, otherwise 0.
2. **`E`-dimensional latent layer**: This is linearly related to the input features. However, multiply each input feature by the epitope weights first, such that the `beta coefficients` for sites not within an epitope are 0.
3. **Custom sigmoid-activated layer**: This transforms each output of the latent layer (`E` in total) by a sigmoid function and attempts to capture a "fraction of antibody binding at epitope `e`". For example, say there are two epitopes, with two fractions `b1` and `b2`. The latent layer will have two corresponding latent phenotypes `l1` and `l2`. Then, `b1` = sigmoid(`l1`) and `b2` = sigmoid(`l2`). Note that `b1` is **independent** of `l2`, and vice versa for `b2` and `l1`. 
4. **Custom 1-dimensional output layer**: This transforms the sigmoid-activated layer into the predicted viral escape fraction. This layer has no weights (not a linear function), and is instead simply a product over each of the fractions (`b1`, `b2`, etc.) in the previous layer. 

## Pseudocode
```python
def EscapeModel(TorchdmsModel):
    """
    a subclass of TorchdmsModel for modeling viral escape

    The layer_sizes and activations do not need to be specified in a model string. They can be fixed like this:
    layer_sizes = [E, E]
    activations = [identity, sigmoid]
    """

    # initialize the model as normal, but add something like:
    self.epitope_weights = epitope_weights
    
    layer_sizes = [epitope_weights.shape[0], epitope_weights.shape[0]]
    activations = ["identity", "sigmoid"]

    def constrain_input(self):
        """
        multiply input features by epitope weights
        """
        return self.input_layer * torch.repeat_interleave(self.epitope_weights, 20)

    def to_latent(self):
        """
        go from input features to latent space
        """
        add_model = nn.Linear(self.input_size, epitope_weights.shape[0], bias = True)
        latent_space = add_model(constrain_input(self))

        return latent_space

    def to_escape_fraction(self):
        """
        go from latent space to the escape fraction
        """
        b_fractions = []
        for e in range(len(epitope_weights)):
            b_fractions.append(sigmoid(to_latent(self)[e]))

        return product(b_fractions)

     def beta_coefficients(self):
         """
         extract beta coefficients for each node in the latent space
         """
         beta_coefficients = []
         for e in range(len(epitope_weights)):
             beta_coefficients.append(self.model.weight.data[e])
  
         return beta_coefficients
```
